src/lib/stringifiers.ts

```diff
@@ -0,0 +1,151 @@
+};
+
+export const stringifyInferred = (node: t.InferredType): string => {
+    return `infer ${node.name}`;
+};
+
+export const stringifyIntersection = (node: t.IntersectionType): string => {
+    return stringifyList(node.types, ' & ');
+};
+
+export const stringifyIntrinsic = (node: t.IntrinsicType): string => {
+    return node.name;
+};
+
+export const stringifyPredicate = (node: t.PredicateType): string => {
+    const out = node.asserts ? ['asserts', node.name] : [node.name];
+    if (node.targetType) {
+        out.push('is', stringifyType(node.targetType));
+    }
+
+    return out.join(' ');
+};
+
+export const stringifyQuery = (node: t.QueryType): string => {
+    return `typeof ${stringifyType(node.queryType)}`;
+};
+
+export const stringifyReference = (node: t.ReferenceType): string => {
+    const name = node.reflection ? node.reflection.name : node.name;
+    let typeArgs = '';
+    if (node.typeArguments) {
+        typeArgs += '<';
+        typeArgs += stringifyList(node.typeArguments, ', ');
+        typeArgs += '>';
+    }
+
+    return name + typeArgs;
+};
+
+export const stringifyReflection = (node: t.ReflectionType): string => {
+    if (!node.declaration.children && node.declaration.signatures) {
+        return 'function';
+    } else {
+        return 'object';
+    }
+};
+
+export const stringifyStringLiteral = (node: t.StringLiteralType): string => {
+    return `"${node.value}"`;
+};
+
+export const stringifyTuple = (node: t.TupleType): string => {
+    return `[${stringifyList(node.elements, ', ')}]`;
+};
+
+export const stringifyTypeOperator = (node: t.TypeOperatorType): string => {
+    return `${node.operator} ${stringifyType(node.target)}`;
+};
+
+export const stringifyTypeParameter = (node: t.TypeParameterType): string => {
+    return node.name;
+};
+
+export const stringifyUnion = (node: t.UnionType): string => {
+    return stringifyList(node.types, ' | ');
+};
+
+export const stringifyUnknown = (node: t.UnknownType): string => {
+    return node.name;
+};
+
+export const stringifyVoid = (): string => {
+    return 'void';
+};
+
+const stringifiers = {
+    array: stringifyArray,
+    conditional: stringifyConditional,
+    indexedAccess: stringifyIndexedAccess,
+    inferred: stringifyInferred,
+    intersection: stringifyIntersection,
+    intrinsic: stringifyIntrinsic,
+    predicate: stringifyPredicate,
+    query: stringifyQuery,
+    reference: stringifyReference,
+    reflection: stringifyReflection,
+    stringLiteral: stringifyStringLiteral,
+    tuple: stringifyTuple,
+    typeOperator: stringifyTypeOperator,
+    typeParameter: stringifyTypeParameter,
+    union: stringifyUnion,
+    unknown: stringifyUnknown,
+    void: stringifyVoid
+};
+
+/**
+ * Return a string representation of the given type.
+ */
+export const stringifyType = (node: t.Type): string => {
+    if (!node || !node.type || !{}.hasOwnProperty.call(stringifiers, node.type)) {
```

Comment:
We control `stringifiers`, no need for the paranoid version of hasOwnProperty. Also, I'd rather not worry about not being provided node/node.type. This is a TS only library with strictNullChecks on, if someone breaks that contract, that's on them.

Reply:
Mostly I did this so that we could better support passing in an object pulled from typedoc JSON, given the lack of deserializing (related: #1180) and my use cases in the RFC. In that case it would probably be better to use an alternative type that ensures we only use properties that exist on both the JSON objects and the live models.